src/os/unix/wayland.rs

```diff
@@ -0,0 +1,1182 @@
+use crate::key_handler::KeyHandler;
+use crate::mouse_handler;
+use crate::rate::UpdateRate;
+use crate::{
+    os::unix::x11::Menu, InputCallback, Key, KeyRepeat, MouseButton, MouseMode, Scale, ScaleMode,
+    WindowOptions,
+};
+use crate::{CursorStyle, MenuHandle, UnixMenu};
+use crate::{Error, Result};
+
+use std::ffi::c_void;
+use std::io::Write;
+use wayland_client::protocol::{
+    wl_buffer::WlBuffer,
+    wl_compositor::WlCompositor,
+    wl_display::WlDisplay,
+    wl_keyboard::KeymapFormat,
+    wl_keyboard::WlKeyboard,
+    wl_pointer::WlPointer,
+    wl_seat::WlSeat,
+    wl_shm::{Format, WlShm},
+    wl_shm_pool::WlShmPool,
+    wl_surface::WlSurface,
+};
+use wayland_client::{Attached, Main};
+use wayland_client::{EventQueue, GlobalManager};
+use wayland_protocols::xdg_shell::client::{
+    xdg_surface::XdgSurface, xdg_toplevel::XdgToplevel, xdg_wm_base::XdgWmBase,
+};
+
+use std::cell::RefCell;
+use std::os::unix::io::RawFd;
+use std::rc::Rc;
+use std::sync::mpsc;
+
+// We have these in C so we can always have optimize on (-O3) so they
+// run fast in debug build as well. These functions should be seen as
+// "system" functions that just doesn't exist in X11
+extern "C" {
+    fn Image_upper_left(
+        target: *mut u32,
+        source: *const u32,
+        source_w: u32,
+        source_h: u32,
+        source_stride: u32,
+        dest_width: u32,
+        dest_height: u32,
+        bg_color: u32,
+    );
+
+    fn Image_center(
+        target: *mut u32,
+        source: *const u32,
+        source_w: u32,
+        source_h: u32,
+        source_stride: u32,
+        dest_width: u32,
+        dest_height: u32,
+        bg_color: u32,
+    );
+
+    fn Image_resize_linear_aspect_fill_c(
+        target: *mut u32,
+        source: *const u32,
+        source_w: u32,
+        source_h: u32,
+        source_stride: u32,
+        dest_width: u32,
+        dest_height: u32,
+        bg_color: u32,
+    );
+
+    fn Image_resize_linear_c(
+        target: *mut u32,
+        source: *const u32,
+        source_w: u32,
+        source_h: u32,
+        source_stride: u32,
+        dest_width: u32,
+        dest_height: u32,
+    );
+}
+
+struct SingleBuffer {
+    fd: std::fs::File,
+    //Shm pool and size
+    pool: (Main<WlShmPool>, i32),
```

Comment:
Is there a reason these are a tuple, or could they be split into:
```suggestion
    pool: Main<WlShmPool>,
    pool_size: i32,
```
?

Reply:
There's actually none.
Is it possible to take a look at these changes locally instead of "on github"?